Update margin box for trimmed block-end boxes in block container and adjust position of self-collapsing children

sammygill opened the pull request:
#### b9ee985deec446b5c3bfb23ea59c90e3c14a814d
<pre>
Update margin box for trimmed block-end boxes in block container and adjust position of self-collapsing children
<a href="https://bugs.webkit.org/show_bug.cgi?id=253679">https://bugs.webkit.org/show_bug.cgi?id=253679</a>
rdar://106524654

Reviewed by Alan Baradlay.

The initial attempt at trimming margins that collapsed through to the
block-end of a block container simply used the MarginInfo for the block
container to &quot;trim,&quot; the margins. This MarginInfo structure held the
margins that collapsed through and would be used with the container&apos;s
actual block-end margin to perform a final margin collapsing. By ignoring
these values we were prohibit the collapsing from occurring with the
block container&apos;s block-end margin.

However, since the margins for the children at the end of the block
container were not actually trimmed (i.e. their m_marginBox still
contained their used margin values), the block container may have
incorrectly positioned self-collapsing children that were also at the
bottom of the block.

container {
    display: block;
    margin-trim: block;
    margin-block-end: 10px;
}
item {
    display: block;
    margin-block-end: 40px;
    inline-size: 50px;
    block-size: 50px;
    background-color: green;
}
.collapsed {
    margin-block-start: 50px;
    block-size: 0px;
}
&lt;/style&gt;
&lt;container&gt;
    &lt;item&gt;&lt;/item&gt;
    &lt;item class=&quot;collapsed&quot;&gt;&lt;/item&gt;
    &lt;item class=&quot;collapsed&quot;&gt;&lt;/item&gt;
&lt;/container&gt;

Here the last two items in the block container are self-collapsing, so
they would normally participate in margin collapsing with the block-end
margin of the first item and the block-end mark of the container itself.
Without margin-trim the collapsed items should have an offset of 100px
from the top of the containing block.

When block-end margin trimming is specified, however, these items will
need to get placed right below the first item since the block-end margin
will be removed from the first item and both block margins will be
trimmed from the last two.

Since we cannot know if a self-collapsing child is at the block-end
of a block container (i.e. there may be another non self-collapsing
child further on in the layout), we need to make these adjustments when
we are handling the after side of the block container. This means we
need to walk back up the children of the container and perform these
adjustments. As we walk up the tree we have 3 scenarios for each child:

1.) The child is a self-collapsing block:
We should trim the before/after margins of this child and make sure its
position is at the bottom of the block right below the last non
self-collapsing child. We do not need to go inside the self-collapsing
children even if they contain more self-collapsing children since those
will already be positioned at the top of their containing block. If they
had any margins they were have collapsed through to the outer most
self-collapsing child.

2.) The child is a nested block
In this case, we will trim its after margin (since it is at the bottom
of the block and everything below it is self-collapsing), but we also
need to check if the margins of its children nested inside can collapse
through with its own after margin. If it cannot (e.g. the nested block
has some border and padding), then we are done since those margins
cannot affect the margins or children outside of their containing block
(the nested block we are checking), and everything before this nested
block cannot affect the bottom of the block.

However, if the margins inside can collapse through, then we need to go
one more level deep (inside of this nested block) and perform the same
logic. This is because those margins could have potentially collapsed
through and affected any self-collapsing children at the bottom of this
nested block as well as the children in the outer one.

3.) The block is some other non-nested and non self-collapsing block.
In this case we can just trim the block-end margin and end our
adjustments since everything before this block cannot impact the bottom
(just like the case with the nested-block that cannot have its children&apos;s
margins collapse through).

In this case, we need to recurse into it and
apply the same type of logic. It is possible that this nested block
container may have its children&apos;s margins at the bottom of the block
contribute to the margin collapsing occurring at the outer block container.

&lt;style&gt;
.outer {
    margin-trim: block;
}
container {
    display: block;
}
item {
    display: block;
    margin-block-end: 40px;
    inline-size: 50px;
    block-size: 50px;
    background-color: green;
}
.collapsed {
    margin-block-start: 50px;
    block-size: 0px;
}
&lt;/style&gt;
&lt;container class=&quot;outer&quot;&gt;
    &lt;container&gt;
        &lt;item data-offset-y=&quot;8&quot;&gt;&lt;/item&gt;
        &lt;item data-offset-y=&quot;58&quot; class=&quot;collapsed&quot;&gt;&lt;/item&gt;
        &lt;item data-offset-y=&quot;58&quot; class=&quot;collapsed&quot;&gt;&lt;/item&gt;
    &lt;/container&gt;
&lt;/container&gt;

Here the outer container has block-end margin trim but there is no
block-end margin set on its last child nor does it have any self
collapsing children at the end of the block. However, it has another
block container nested within its last child. This nested block
container has self-collapsing children at the end of the block that
have their margins collapse through and propagate up to the outer
container.

* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-last-child-with-border-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-last-child-with-border.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-nested-last-child-with-border-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-nested-last-child-with-border.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-nested-at-bottom-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-nested-at-bottom.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-nested-margin-trim-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-nested-margin-trim.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-offsets-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-offsets-nested-multiple-times-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-offsets-nested-multiple-times.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-offsets-nested-once-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-offsets-nested-once.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-offsets-vert-lr-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-offsets-vert-lr.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-offsets.html: Added.
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::layoutBlockChildren):
(WebCore::RenderBlockFlow::trimBlockEndChildrenMargins):
* Source/WebCore/rendering/RenderBlockFlow.h:

Canonical link: <a href="https://commits.webkit.org/261750@main">https://commits.webkit.org/261750@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2b1e22ab16740cd3e52283b902cc3ccfc4222650

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112655 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21808 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1324 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4430 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121184 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116722 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23149 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12973 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/5582 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118423 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/17194 "4 failures") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100417 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105730 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99145 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/963 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46215 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14134 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/1000 "1 flakes 6 failures") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/95425 "275 api tests failed or timed out") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14818 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/10359 "1 flakes 2 failures") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20151 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52996 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8200 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16662 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -818,7 +818,32 @@ void RenderBlockFlow::layoutBlockChildren(bool relayoutChildren, LayoutUnit& max
+
+    // If we are trimming the block end margin, we need to make sure we actually trim the margin of the children
+    // at the end of the block by walking back up the container. Any self collapsing children will also need to
+    // have their position adjusted to below the last non self-collapsing child.
```
Are you planning on transitioning this to handle descendant block boxes with collapsed through margins? Within a BFC, you can have arbitrary number of block boxes at arbitrary nesting levels.

sammygill replied:
Yes you are right. I think I will try to address that in a separate patch since I can see that being slightly more complicated (unless you think both scenarios should be addressed in a single patch). I'll make the necessary changes to the patch/bug report information

alanbaradlay replied:
While this patch fixes these test included cases, real web content tend to be slightly more complicated with block content nesting. However I am ok with landing this patch as is, if we need to quickly address a broken popular content.  Now if this is just regular fix, I'd try to figure out first how nested cases would be handled as they may require a very different approach e.g. instead of back-tracing you may be building a list of block boxes that touch the block-end side of the BFC root as block layout progresses -as by definition the in-flow block level box that is being laid out is adjacent to the BFC root's content box (bottom) (under certain conditions).

sammygill replied:
Modified the patch to try and detect this type of nesting and recurse into the nested block container and perform the same logic. I think the number of scenarios in which type of nesting can occur and have an impact on the block-end margins is more limited than I initially thought there were going to be so I provided my reasoning in the writeup with an example, but correct me if I'm wrong!

alanbaradlay replied:
> I think the number of scenarios in which type of nesting can occur and have an impact on the block-end margins is more limited than I initially thought

You surely thought about it a lot more than I did, so I am curious of what scenarios you had in your mind (that turned out to be non-issues). Anyway when I made that comment I was just thinking along the lines of arbitrary number of block level boxes with collapsed through margins at arbitrary level of nesting....which may make you consider a different approach to ensure performant margin-trimming (back-tracing vs. tracking boxes with collapsed through margins).
It looks like you decided to continue with back-tracing.
> This means we need to walk back up the children of the container and perform these  adjustments until we hit a child that is not self-collapsing

Is that because you think authors don't create such arbitrarily expensive content or some other reasons? (I personally don't think they do, but I also don't have any data to back up such claims).  
If we stick with this approach I think the preferred way of doing it is a non-recursive style.

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -819,11 +819,47 @@ void RenderBlockFlow::layoutBlockChildren(bool relayoutChildren, LayoutUnit& max
+    // have their position adjusted to below the last non self-collapsing child in its containing block
+    auto* child = lastChildBox();
+    while (child) {
+        if (child->isExcludedFromNormalLayout() || child->isFloatingOrOutOfFlowPositioned()) {
```
I think checking against !inFlow()(instead of isFloatingOrOut....) is more future proof.

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -819,11 +819,47 @@ void RenderBlockFlow::layoutBlockChildren(bool relayoutChildren, LayoutUnit& max
+            childContainingBlock->setMarginBeforeForChild(*child, 0_lu);
+            childContainingBlock->setLogicalTopForChild(*child, childContainingBlock->logicalHeight());
+            child = child->previousSiblingBox();
+        }  else if (auto *nestedBlock = dynamicDowncast<RenderBlockFlow>(child); nestedBlock && nestedBlock->isBlockContainer() && !nestedBlock->childrenInline() && !nestedBlock->style().marginTrim().contains(MarginTrimType::BlockEnd)) {
```
if (auto* nestedBlock

sammygill committed:
Update margin box for trimmed block-end boxes in block container and adjust position of self-collapsing children
https://bugs.webkit.org/show_bug.cgi?id=253679
rdar://106524654

Reviewed by Alan Baradlay.

The initial attempt at trimming margins that collapsed through to the
block-end of a block container simply used the MarginInfo for the block
container to "trim," the margins. This MarginInfo structure held the
margins that collapsed through and would be used with the container's
actual block-end margin to perform a final margin collapsing. By ignoring
these values we were prohibit the collapsing from occurring with the
block container's block-end margin.

However, since the margins for the children at the end of the block
container were not actually trimmed (i.e. their m_marginBox still
contained their used margin values), the block container may have
incorrectly positioned self-collapsing children that were also at the
bottom of the block.

container {
    display: block;
    margin-trim: block;
    margin-block-end: 10px;
}
item {
    display: block;
    margin-block-end: 40px;
    inline-size: 50px;
    block-size: 50px;
    background-color: green;
}
.collapsed {
    margin-block-start: 50px;
    block-size: 0px;
}
</style>
<container>
    <item></item>
    <item class="collapsed"></item>
    <item class="collapsed"></item>
</container>

Here the last two items in the block container are self-collapsing, so
they would normally participate in margin collapsing with the block-end
margin of the first item and the block-end mark of the container itself.
Without margin-trim the collapsed items should have an offset of 100px
from the top of the containing block.

When block-end margin trimming is specified, however, these items will
need to get placed right below the first item since the block-end margin
will be removed from the first item and both block margins will be
trimmed from the last two.

Since we cannot know if a self-collapsing child is at the block-end
of a block container (i.e. there may be another non self-collapsing
child further on in the layout), we need to make these adjustments when
we are handling the after side of the block container. This means we
need to walk back up the children of the container and perform these
adjustments. As we walk up the tree we have 3 scenarios for each child:

1.) The child is a self-collapsing block:
We should trim the before/after margins of this child and make sure its
position is at the bottom of the block right below the last non
self-collapsing child. We do not need to go inside the self-collapsing
children even if they contain more self-collapsing children since those
will already be positioned at the top of their containing block. If they
had any margins they were have collapsed through to the outer most
self-collapsing child.

2.) The child is a nested block
In this case, we will trim its after margin (since it is at the bottom
of the block and everything below it is self-collapsing), but we also
need to check if the margins of its children nested inside can collapse
through with its own after margin. If it cannot (e.g. the nested block
has some border and padding), then we are done since those margins
cannot affect the margins or children outside of their containing block
(the nested block we are checking), and everything before this nested
block cannot affect the bottom of the block.

However, if the margins inside can collapse through, then we need to go
one more level deep (inside of this nested block) and perform the same
logic. This is because those margins could have potentially collapsed
through and affected any self-collapsing children at the bottom of this
nested block as well as the children in the outer one.

3.) The block is some other non-nested and non self-collapsing block.
In this case we can just trim the block-end margin and end our
adjustments since everything before this block cannot impact the bottom
(just like the case with the nested-block that cannot have its children's
margins collapse through).

In this case, we need to recurse into it and
apply the same type of logic. It is possible that this nested block
container may have its children's margins at the bottom of the block
contribute to the margin collapsing occurring at the outer block container.

<style>
.outer {
    margin-trim: block;
}
container {
    display: block;
}
item {
    display: block;
    margin-block-end: 40px;
    inline-size: 50px;
    block-size: 50px;
    background-color: green;
}
.collapsed {
    margin-block-start: 50px;
    block-size: 0px;
}
</style>
<container class="outer">
    <container>
        <item data-offset-y="8"></item>
        <item data-offset-y="58" class="collapsed"></item>
        <item data-offset-y="58" class="collapsed"></item>
    </container>
</container>

Here the outer container has block-end margin trim but there is no
block-end margin set on its last child nor does it have any self
collapsing children at the end of the block. However, it has another
block container nested within its last child. This nested block
container has self-collapsing children at the end of the block that
have their margins collapse through and propagate up to the outer
container.

* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-last-child-with-border-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-last-child-with-border.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-nested-last-child-with-border-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-nested-last-child-with-border.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-nested-at-bottom-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-nested-at-bottom.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-nested-margin-trim-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-nested-margin-trim.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-offsets-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-offsets-nested-multiple-times-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-offsets-nested-multiple-times.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-offsets-nested-once-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-offsets-nested-once.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-offsets-vert-lr-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-offsets-vert-lr.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-box/margin-trim/block-container-block-end-self-collapsing-children-offsets.html: Added.
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::layoutBlockChildren):
(WebCore::RenderBlockFlow::trimBlockEndChildrenMargins):
* Source/WebCore/rendering/RenderBlockFlow.h:

Canonical link: https://commits.webkit.org/261750@main
